WooCommerce/Classes/ViewRelated/Blaze/AdDestination/BlazeAddParameterViewModel.swift

```diff
@@ -0,0 +1,76 @@
+import Foundation
+
+/// View model for `BlazeAddParameterView`
+final class BlazeAddParameterViewModel: ObservableObject {
+    @Published var key: String
+    @Published var value: String
+    @Published var hasValidationError: Bool = false
+    @Published var hasCountError: Bool = false
+
+    let remainingCharacters: Int
+    let isNotFirstParameter: Bool
+    let parameter: BlazeAdURLParameter?
+
+    // For adding or editing a new parameter, the two inputs are to be combined to be "key=value".
+    // However, for adding or editing 2nd or more parameters, the input becomes "&key=value" due to how URL parameter work.
+    var totalInputLength: Int {
+        let inputString = (isNotFirstParameter ? "&" : "") + key + "=" + value
+        return inputString.count
+    }
+
+    var shouldDisableSaveButton: Bool {
+        key.isEmpty || value.isEmpty || hasCountError || hasValidationError
+    }
+
+    typealias BlazeAddParameterCompletionHandler = (_ key: String, _ value: String) -> Void
+    private let completionHandler: BlazeAddParameterCompletionHandler
+
+    private let cancellationHandler: () -> Void
+
+    init(remainingCharacters: Int,
+         isNotFirstParameter: Bool = true,
+         parameter: BlazeAdURLParameter? = nil,
+         onCancel: @escaping () -> Void,
+         onCompletion: @escaping BlazeAddParameterCompletionHandler) {
+        self.remainingCharacters = remainingCharacters
+        self.isNotFirstParameter = isNotFirstParameter
+        self.parameter = parameter
+        self.cancellationHandler = onCancel
+        self.completionHandler = onCompletion
+
+        key = parameter?.key ?? ""
+        value = parameter?.value ?? ""
+    }
+
+    func didTapCancel() {
+        cancellationHandler()
+    }
+
+    func didTapSave() {
+        completionHandler(key, value)
+    }
+
+    func validateInput(text: String) {
+        validateParameter(text: text)
+        validateInputLength()
+    }
+
+    /// This function validates the URL parameters using String.isValidURL().
+    /// It requires a full URL, thus Constant.baseURLForValidation is added.
+    /// The constant uses "?key=", but it's also used to validate parameter keys, as they both follow the same validation rule.
+    ///
+    private func validateParameter(text: String) {
+        let url = Constant.baseURLForValidation + text
+        hasValidationError = !url.isValidURL()
+    }
```

Comment:
Here you only validate either key or value at a time. I found an edge case where the initial error is bypassed:
1. Type a key with space - the validation error is displayed.
2. Switch to the value field and enter a valid content - the validation error disappears.
3. Tap save, you get invalid key and valid field.

To cover this case, you should validate both at the same time:
```suggestion
    private func validateParameter() {
        let url = "https://woo.com/?\(key)=\(value)" // use constant for this
        hasValidationError = !url.isValidURL()
    }
```

Reply:
That's a great discovery, and it makes a lot more sense to validate all at once all the time. I committed the suggestion and added https://github.com/woocommerce/woocommerce-ios/pull/11642/commits/7909e8abafe93d3176f9bdafba5893d74834f5ce